src/js/tls.js

```diff
@@ -48,51 +50,40 @@ function TLSSocket(socket, options) {
 
   native.TlsInit(this, options, secureContext);
   this._socketState = socket._socketState;
+
+  var self = this;
+  process.nextTick(function() {
+    if (socket._writableState.ready && !options.isServer) {
+      self._native_connect(options.servername || options.host || 'localhost');
+      self._native_read(null);
+    }
+  });
 }
-util.inherits(TLSSocket, EventEmitter);
+util.inherits(TLSSocket, Duplex);
 
-TLSSocket.prototype._read = native.read;
-TLSSocket.prototype._write = native.write;
-TLSSocket.prototype._connect = native.connect;
+TLSSocket.prototype._native_read = native.read;
+TLSSocket.prototype._native_write = native.write;
+TLSSocket.prototype._native_connect = native.connect;
 
 TLSSocket.prototype.connect = function(options, callback) {
-  this._connect(options.servername || options.host || 'localhost');
-
   if (util.isFunction(callback)) {
     this.on('secureConnect', callback);
   }
-
-  this._socket.connect(options);
-};
-
-TLSSocket.prototype.write = function(data, callback) {
-  if (!Buffer.isBuffer(data)) {
-    data = new Buffer(data);
+  if (this._socket instanceof net.Socket) {
+    this._native_connect(options.servername || options.host || 'localhost');
+    this._socket.connect(options);
   }
-
-  data = this._write(data);
-  return this._socket.write(data, callback);
 };
 
-TLSSocket.prototype.pause = function() {
-  this._socket.pause();
-};
-
-TLSSocket.prototype.resume = function() {
-  this._socket.resume();
+TLSSocket.prototype._write = function(chunk, callback, onwrite) {
+  chunk = this._native_write(chunk);
+  this._socket.write(chunk, callback);
+  onwrite();
 };
 
 TLSSocket.prototype.end = function(data, callback) {
-  if (data) {
-    if (!Buffer.isBuffer(data)) {
-      data = new Buffer(data);
-    }
-    data = this._write(data, true);
-  } else {
-    data = this._write(null, true);
-  }
-
-  this._socket.end(data, callback);
+  Writable.prototype.end.call(this, data, callback);
```

Comment:
I think Duplex.prototype.end would work here as well, and we could remove the Writable requirement.